crypto/rand/drbg_hash.c

```diff
@@ -0,0 +1,292 @@
+ */
+static int hash_df(RAND_DRBG *drbg, unsigned char *out,
+                   const unsigned char inbyte,
+                   const unsigned char *in, size_t inlen,
+                   const unsigned char *in2, size_t in2len,
+                   const unsigned char *in3, size_t in3len)
+{
+    RAND_DRBG_HASH *hash = &drbg->data.hash;
+    EVP_MD_CTX *ctx = hash->ctx;
+    unsigned char *vtmp = hash->vtmp;
+    unsigned char tmp[1+4+1];
+    unsigned char tmp_sz = 5;
+    /* Standard only ever needs seedlen bytes which is always less than
+     * maximum permitted so no need to check length.
+     */
+    size_t outlen = drbg->seedlen;
+    size_t num_bits_returned = outlen * 8;
+    /* (Step 3) counter=1 tmp[0] is the 8 bit counter. */
+    tmp[0] = 1;
+    /* tmp[1..4] is the fixed 32 bit no_of_bits_to_return */
+    tmp[1] = (num_bits_returned >> 24) & 0xff;
+    tmp[2] = (num_bits_returned >> 16) & 0xff;
+    tmp[3] = (num_bits_returned >> 8) & 0xff;
+    tmp[4] = num_bits_returned & 0xff;
+    /* Tack the additional input byte onto the end of tmp if it exists */
+    if (inbyte != INBYTE_IGNORE)
+        tmp[tmp_sz++] = inbyte;
+
+    /* (Step 4) */
+    for (;;) {
+        /* (Step 4.1) temp = temp || Hash(counter||num_bits_returned||input */
+        if (!(EVP_DigestInit_ex(ctx, hash->md, NULL)
+                && EVP_DigestUpdate(ctx, tmp, tmp_sz)
+                && EVP_DigestUpdate(ctx, in, inlen)
+                && (in2 == NULL || EVP_DigestUpdate(ctx, in2, in2len))
+                && (in3 == NULL || EVP_DigestUpdate(ctx, in3, in3len))))
+            return 0;
+
+        if (outlen < hash->blocklen) {
+            if (!EVP_DigestFinal(ctx, vtmp, NULL))
+                return 0;
+            memcpy(out, vtmp, outlen);
+            OPENSSL_cleanse(vtmp, hash->blocklen);
+            break;
+        }
+        else if(!EVP_DigestFinal(ctx, out, NULL))
+            return 0;
+
+        outlen -= hash->blocklen;
+        if (outlen == 0)
+            break;
+        /* (Step 4.2) counter++ */
+        tmp[0]++;
+        out += hash->blocklen;
+    }
+    return 1;
+}
+
+static int hash_df1(RAND_DRBG *drbg, unsigned char *out,
+                    const unsigned char in_byte,
+                    const unsigned char *in1, size_t in1len)
+{
+    return hash_df(drbg, out, in_byte, in1, in1len, NULL, 0, NULL, 0);
+}
+
+/*
+ * Add 2 byte buffers together. The first elements in each buffer are the top
+ * most bytes. The result is stored in the dst buffer.
+ * The final carry is ignored i.e: dst =  (dst + in) mod (2^seedlen_bits).
+ * where dst size is drbg->seedlen, and inlen <= drbg->seedlen.
+ */
+static int add_bytes(RAND_DRBG *drbg, unsigned char *dst,
+                     unsigned char *in, size_t inlen)
+{
+    size_t i;
+    int result;
+    const unsigned char *add;
+    unsigned char carry = 0, *d;
+
+    if (drbg->seedlen < 1 || inlen < 1 || inlen > drbg->seedlen)
```

Comment:
Is this a "should never happen" scenario? In which case would `ossl_assert` be more appropriate?